backport from Emacs 23 - and fix for pop-to-buffer to do set input focus in new frames
Index: src/buffer.c
===================================================================
RCS file: /sources/emacs/emacs/src/buffer.c,v
retrieving revision 1.526.2.5
diff -c -r1.526.2.5 buffer.c
*** src/buffer.c	8 Jan 2008 04:30:20 -0000	1.526.2.5
--- src/buffer.c	19 Dec 2008 12:38:04 -0000
***************
*** 1713,1718 ****
--- 1713,1721 ----
  do not put this buffer at the front of the list of recently selected ones.
  This function returns the buffer it switched to.
  
+ If the selected window is the minibuffer window or dedicated to
+ its buffer, use `pop-to-buffer' for displaying the buffer.
+ 
  WARNING: This is NOT the way to work on another buffer temporarily
  within a Lisp program!  Use `set-buffer' instead.  That avoids messing with
  the window-buffer correspondences.  */)
***************
*** 1735,1744 ****
        return Fset_buffer (buffer);
      }
  
!   err = no_switch_window (selected_window);
!   if (err) error (err);
! 
!   return switch_to_buffer_1 (buffer, norecord);
  }
  
  DEFUN ("pop-to-buffer", Fpop_to_buffer, Spop_to_buffer, 1, 3, 0,
--- 1738,1750 ----
        return Fset_buffer (buffer);
      }
  
!   if (EQ (minibuf_window, selected_window)
!       || !NILP (Fwindow_dedicated_p (selected_window)))
!     /* We can't use the selected window so let `pop-to-buffer' try some
!        other window. */
!     return call3 (intern ("pop-to-buffer"), buffer, Qnil, norecord);
!   else
!     return switch_to_buffer_1 (buffer, norecord);
  }
  
  DEFUN ("pop-to-buffer", Fpop_to_buffer, Spop_to_buffer, 1, 3, 0,
*** src/buffer.c	25 Feb 2006 23:33:57 +0000	1.501
--- src/buffer.c	07 May 2006 22:45:45 +0100	
***************
*** 175,180 ****
--- 175,182 ----
  Lisp_Object Qinsert_in_front_hooks;
  Lisp_Object Qinsert_behind_hooks;
  
+ Lisp_Object Qselect_frame_set_input_focus;
+ 
  static void alloc_buffer_text P_ ((struct buffer *, size_t));
  static void free_buffer_text P_ ((struct buffer *b));
  static struct Lisp_Overlay * copy_overlays P_ ((struct buffer *, struct Lisp_Overlay *));
***************
*** 1723,1729 ****
  	}
      }
    Fset_buffer (buf);
!   Fselect_window (Fdisplay_buffer (buf, other_window, Qnil), norecord);
    return buf;
  }
  
--- 1725,1734 ----
  	}
      }
    Fset_buffer (buf);
!   call1(Qselect_frame_set_input_focus,
! 	Fwindow_frame( Fselect_window (Fdisplay_buffer (buf, 
! 							other_window, Qnil), 
! 				       norecord)));
    return buf;
  }
  
***************
*** 5227,5232 ****
--- 5232,5239 ----
    Qafter_change_functions = intern ("after-change-functions");
    staticpro (&Qafter_change_functions);
    staticpro (&Qucs_set_table_for_input);
+   Qselect_frame_set_input_focus = intern ("select-frame-set-input-focus");
+   staticpro (&Qselect_frame_set_input_focus);
  
    Qkill_buffer_query_functions = intern ("kill-buffer-query-functions");
    staticpro (&Qkill_buffer_query_functions);
